napari/layers/utils/layer_utils.py

```diff
@@ -204,81 +216,485 @@ def _nanmax(array):
+        maximum number of chunks per plane.
+    max_allowed_chunks: None | int
+        The total number of chunks that is allowed based on the chunk size and the PIXEL_TRESHOLD. None if the
+        chunk size is higher than the PIXEL_THRESHOLD.
+    chunk_size_y: int
+        Size of the slice of the y dimension if dealing with numpy array, otherwise the y dim size of the chunk when
+        dealing with chunked array data.
+    chunk_size_x: int
+        Size of the slice of the x dimension if dealing with numpy array, otherwise the x dim size of the chunk when
+        dealing with chunked array data.
+    """
+    # defaults in case we are dealing with numpy array in which case chunks are crops and chunk size is just size of the
+    # slice. If chunked array these can be overwritten.
+    max_chunks_per_plane = 9
+    max_allowed_chunks = len(plane_indices) * max_chunks_per_plane
+    chunk_size_y = chunk_size_x = int(
+        (PIXEL_THRESHOLD // max_allowed_chunks) ** 0.5
+    )
+    if chunk_shape:
+        chunk_size_product = int(np.prod(chunk_shape))
+        chunk_plane_size = chunk_shape[-offset:]
+        chunk_size_y, chunk_size_x = chunk_plane_size[0], chunk_plane_size[1]
+        max_allowed_chunks = PIXEL_THRESHOLD // chunk_size_product
+        # in case of the chunk size going over the pixel threshold, we can wait until data is in memory.
+        if max_allowed_chunks and len(plane_indices) != 0:
+            max_chunks_per_plane = max_allowed_chunks // len(plane_indices)
+
+    return max_chunks_per_plane, max_allowed_chunks, chunk_size_y, chunk_size_x
+
+
+def _get_pixel_start_indices(
+    plane_indices: Sequence[Sequence[int]],
+    plane_shape,
+    offset: int,
+    chunk_size_y: int,
+    chunk_size_x: int,
+    chunk_shape: None | Tuple[int, ...] = None,
+) -> Tuple[Sequence[Sequence[int]], List[int], List[int]]:
+    """
+    Get the start indices of the individual planes and the y and x dimension in pixel space.
+
+    In case of chunked array data, the plane_indices are converted to "voxel space". For y and x start indices, if
+    dealing with chunked array data the index refers to the actual chunk and not pixel. In case of dealing with
+    numpy array it refers to the starting pixels for each slice.
+
+    Parameters
+    ----------
+    plane_indices: Sequence[Sequence[int]]
+        Bottom, middle and top plane or single plane index for each non-visible dimension.
+    plane_shape
+    offset: int
+        Number of visible dimensions.
+    chunk_shape: None | Tuple[int]
+        The size per dimension of the chunks.
+    chunk_size_y: int
+        Size of the slice of the y dimension if dealing with numpy array, otherwise the y dim size of the chunk when
+        dealing with chunked array data.
+    chunk_size_x: int
+        Size of the slice of the x dimension if dealing with numpy array, otherwise the x dim size of the chunk when
```

Comment:
why we care about what it means for numpy array, when there is no chunk on numpy array?

Reply:
In order not to make the functions longer in case of numpy it is the length of the slice. So later on if we define our start point we can just say `+ chunk_size_y` which in case of chunks can be multiplied and in case of `numpy.array` is just used directly.